### PR TITLE
Updates base images and implements affiliate fee.

### PR DESCRIPTION
**Pull request by MicahZoltu**
Also adds defaultGasPrice both to facilitate testing and because ethers was defaulting to 20 nanoeth.

Sadly, the integration tests now take ~45 seconds to run against Geth, and that is with not cleaning up between tests (only cleaning up at beginning/end).

### CODE REVIEW

File: toolbox/libraries/liquid-long-ethers-impl.ts [hunk 1]
```diff
@@ -29,8 +24,8 @@ export class LiquidLongDependenciesEthers implements Dependencies<ethers.utils.B
 		// https://github.com/ethers-io/ethers.js/issues/321
 		const gasEstimate = (await this.provider.estimateGas(transaction)).toNumber()
 		const gasLimit = Math.min(Math.max(Math.round(gasEstimate * 1.3), 250000), 5000000)
-		// TODO: figure out a way to propagate a warning up to the user in this scenario, we don't currently have a mechanism for error propagation, so will require infrastructure work
-		transaction = Object.assign({}, transaction, { gasLimit: gasLimit })
+		const gasPrice = ethers.utils.bigNumberify(this.gasPriceInNanoeth * 1e9)
```
Comment (epheph) [inline toolbox/libraries/liquid-long-ethers-impl.ts]: This dropped a TODO which is still a valid TODO; when an estimate is going to be 6M, we leave it at 5M then proceed to sign TX, which might not have enough gas

File: toolbox/libraries/liquid-long-ethers-impl.ts [hunk 2]
```diff
@@ -13,12 +13,7 @@ export interface Provider {
 }
 
 export class LiquidLongDependenciesEthers implements Dependencies<ethers.utils.BigNumber> {
-	private readonly provider: Provider
-	private readonly wallet: Wallet
-	public constructor(provider: Provider, wallet: Wallet, gasPriceInNeth: number) {
-		this.provider = provider
-		this.wallet = wallet
-	}
+	public constructor(private readonly provider: Provider, private readonly wallet: Wallet, private readonly gasPriceInNanoeth: number) {}
```
Comment (epheph) [inline toolbox/libraries/liquid-long-ethers-impl.ts]: In the future, we should make getGasPrice a function that return a promise?

File: contracts/source/liquid-long.sol [hunk 3]
```diff
@@ -250,7 +250,6 @@ contract PullPayment {
 		address payee = msg.sender;
 		uint256 payment = payments[payee];
 
-		require(payment != 0);
```
Comment (epheph) [inline contracts/source/liquid-long.sol]: Since we are locally modifying solidity we are attributing, perhaps comment it with a line explaining it?

File: client-library/integration-tests/source/integration.ts [hunk 4]
```diff
@@ -3,85 +3,186 @@ require('source-map-support').install()
 import 'mocha'
 import { expect } from 'chai'
 import { LiquidLong } from '@keydonix/liquid-long-client-library'
+import { TimeoutScheduler } from '@keydonix/liquid-long-client-library/output/scheduler';
```
Comment (epheph) [inline client-library/integration-tests/source/integration.ts]: is it odd to pull from `/output` ? 🤷‍♂️

Comment (epheph) [reply client-library/integration-tests/source/integration.ts]: (is it due to `file:./` ?)

File: client-library/integration-tests/source/integration.ts [hunk 5]
```diff
@@ -3,85 +3,186 @@ require('source-map-support').install()
...
 	before(async () => {
 		ethereumAddress = getEnv('ETHEREUM_HTTP', 'http://localhost:1235')
-		liquidLongAddress = getEnv('ETHEREUM_LIQUID_LONG_ADDRESS', '0x80F8DAa435a9AB4B1802Ba56FE7e0abd0f8aB3D3')
+		privateKey = getEnv('ETHEREUM_PRIVATE_KEY', 'fae42052f82bed612a724fec3632f325f377120592c75bb78adfcceae6470c5a')
+		liquidLongAddress = getEnv('ETHEREUM_LIQUID_LONG_ADDRESS', '0xB03CF72BC5A9A344AAC43534D664917927367487')
 		oasisAddress = getEnv('ETHEREUM_OASIS_ADDRESS', '0x3c6721551c2ba3973560aef3e11d34ce05db4047')
 		wethAddress = getEnv('ETHEREUM_WETH_ADRESS', '0xfcaf25bf38e7c86612a25ff18cb8e09ab07c9885')
 		makerAddress = getEnv('ETHEREUM_TUB_ADRESS', '0x93943fb2d02ce1101dadc3ab1bc3cab723fd19d6')
 		daiAddress = getEnv('ETHEREUM_DAI_ADDRESS', '0x8c915bd2c0df8ba79a7d28538500a97bd15ea985')
 		wethAddress = getEnv('ETHEREUM_WETH_ADDRESS', '0xfcaf25bf38e7c86612a25ff18cb8e09ab07c9885')
```
Comment (epheph) [inline client-library/integration-tests/source/integration.ts]: weth address appears twice. Also, a lot of these this should probably just come off the liquidLong contract. Not part of this PR, not a big deal

File: client-library/library/source/liquid-long.ts [hunk 6]
```diff
@@ -142,6 +154,12 @@ export class LiquidLong {
 		const loanInEth = ethLockedInCdp - leverageSizeInEth
 		return loanInEth
 	}
+
+	static validateAndNormalizeAffiliateAddress = (affiliateAddress: string): string => {
+		const match = /^(?:0x)?([a-zA-Z0-9]{40})$/.exec(affiliateAddress)
```
Comment (epheph) [inline client-library/library/source/liquid-long.ts]: `a-fA-F` ?

File: toolbox/libraries/liquid-long-ethers-impl.ts (hunk 2 above)
Comment (MicahZoltu) [reply toolbox/libraries/liquid-long-ethers-impl.ts]: It isn't part of `Dependencies`, which is where the "everything must be async" requirement comes from.  In this case, this implementation of `Dependencies` accepts a configurable `gasPrice` for its execution, and for this particular implementation we use a hard-coded value.

That being said, we do the same thing for the library implementation, can you think of a reason we would want that value to be dynamic?  Perhaps you are thinking that the function should supply a `getGasPriceInNanoeth` function so that the library user can choose to do something more complex for gas pricing (like fetching from an external API)?

Comment (MicahZoltu) [reply toolbox/libraries/liquid-long-ethers-impl.ts]: I went ahead and changed this to a getter in the library (but not here).

File: client-library/integration-tests/source/integration.ts (hunk 5 above)
Comment (MicahZoltu) [reply client-library/integration-tests/source/integration.ts]: None of them are currently exposed by the library.  We could expose them, but I have been trying to keep the library trim (already I have added several "admin" functions).

File: client-library/integration-tests/source/integration.ts [hunk 7]
```diff
@@ -3,85 +3,186 @@ require('source-map-support').install()
...
+			liquidLong.owner.adminWithdrawFees(),
+			liquidLong.affiliate.adminWithdrawFees(),
+		])
+		await (await wallets.owner.sendTransaction({ gasLimit: 21000, gasPrice: 0, to: wallets.funder.address, value: await wallets.owner.getBalance() })).wait()
+		await (await wallets.user.sendTransaction({ gasLimit: 21000, gasPrice: 0, to: wallets.funder.address, value: await wallets.user.getBalance() })).wait()
+		await (await wallets.affiliate.sendTransaction({ gasLimit: 21000, gasPrice: 0, to: wallets.funder.address, value: await wallets.affiliate.getBalance() })).wait()
+	}
+
+	const clearOasisOrderbook = async () => {
+		while (true) {
```
[the remark this answers is not in the document]
Comment (MicahZoltu) [reply client-library/integration-tests/source/integration.ts]: Went ahead and implemented the nonce tracking stuff and parallelized this, we could probably parallelize elsewhere as well later.

[the remark this answers is not in the document]
Comment (MicahZoltu) [reply client-library/integration-tests/source/integration.ts]: Somehow, this code is racy (no idea how).  I'm using the same instance of the class everywhere, yet sometimes it will use the same nonce twice (which shouldn't be possible).  Going to remove it for now, let me know if you see the bug.
```
this.lastUsedNonce = Math.max(this.lastUsedNonce + 1, await this.provider.getTransactionCount(await this.signer.getAddress()))
```
```
Error: Transaction nonce is too low. Try incrementing the nonce.
```
For now, going to revert

[the remark this answers is not in the document]
Comment (MicahZoltu) [reply client-library/integration-tests/source/integration.ts]:
```typescript
const clearWethForDaiBook = async () => {
	while (true) {
		const bestOfferId = await oasis.getBestOffer_(wethAddress, daiAddress)
		if (bestOfferId.isZero()) break
		await oasis.cancel(bestOfferId)
	}
}
const clearDaiForWethBook = async () => {
	while (true) {
		const bestOfferId = await oasis.getBestOffer_(daiAddress, wethAddress)
		if (bestOfferId.isZero()) break
		await oasis.cancel(bestOfferId)
	}
}

await Promise.all([
	clearWethForDaiBook(),
	clearDaiForWethBook(),
])
```

[the remark this answers is not in the document]
Comment (MicahZoltu) [reply client-library/integration-tests/source/integration.ts]: I also verified that the ethers.js noncer wasn't being used.

File: client-library/integration-tests/source/integration.ts (hunk 4 above)
Comment (MicahZoltu) [reply client-library/integration-tests/source/integration.ts]: It is because we are reaching into un-exposed parts of the library.  This was done to help with testing, you should never do this in production.

File: client-library/integration-tests/source/integration.ts (hunk 5 above)
Comment (epheph) [reply client-library/integration-tests/source/integration.ts]: Got it, i was confusing the wrapped abi with the library itself

File: client-library/integration-tests/source/integration.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (epheph) [reply client-library/integration-tests/source/integration.ts]: `this.lastUsedNonce = Math.max(this.lastUsedNonce + 1, await this.provider.getTransactionCount(await this.signer.getAddress()))`

You're awaiting inside something that should be a transaction or unyielding. Even if the first argument of Math.max is the one used, it yields to the event loop and someone else can come along and try again, getting the same answer.

File: toolbox/libraries/liquid-long-ethers-impl.ts (hunk 2 above)
Comment (epheph) [reply toolbox/libraries/liquid-long-ethers-impl.ts]: Ya exactly, hitting some oracle or something. Definitely not needed for MVP, just future consideration; sounds like the infra is set up now to easily make that change later 👍
